ModelHub[Athena]: Support FunnelDiscovery Model

KathiaBarahona opened the pull request:
No changes in modelhub, but seems comparing float series with nan was failing.

thijs-obj reviewed (changes requested):
I am not a fan of this change.
**Question:** does this fix a specific problem that we cannot fix otherwise?

### Reasoning 
#### 1 - Not pandas compatible
Without this change, the behaviour on BigQuery and Athena is the same as in pandas (if I am not mistaken). It is in line with what IEEE 754 says [1], and afaik that is the standard that virtually all programming languages and architectures follow.

The big problem with default behaviour seems to be Postgres, as it has non-standard behaviour [2]. But I don't really see a quick work around for that :( Which is a pitty, because having consistent behaviour would be really nice. Postgres is just sort of broken in that respect (although for a reason).

#### 2 - Different behaviour for literals and Series
The special case in `__eq__` only considers when `other` is a python float with value `nan`, but if `other` is a Series that has the value `nan`, then the default behaviour happens. I think it's weird that the same comparison then gives a different result.

```python
df['nan'] = float('nan')
df['eq_nan1'] = df['a'] == float('nan')
df['eq_nan2'] = df['a'] == df['nan']
# now 'eq_nan1' and 'eq_nan2' should have the same value, but they can be different
```

This is not a super big deal, but not ideal.

### References
[1] https://en.wikipedia.org/wiki/NaN#Comparison_with_NaN - NaN behaviour in IEEE 754 according to wikipedia, I didn't read the actual standard of course :)

[2] https://www.postgresql.org/docs/current/datatype-numeric.html#:~:text=power%20yields%20one.)-,Note,NaN%20values%20as%20equal%2C%20and%20greater%20than%20all%20non%2DNaN%20values.,-The%20types%20decimal

KathiaBarahona commented:
> I am not a fan of this change. **Question:** does this fix a specific problem that we cannot fix otherwise?
> 
> ### Reasoning
> #### 1 - Not pandas compatible
> Without this change, the behaviour on BigQuery and Athena is the same as in pandas (if I am not mistaken). It is in line with what IEEE 754 says [1], and afaik that is the standard that virtually all programming languages and architectures follow.
> 
> The big problem with default behaviour seems to be Postgres, as it has non-standard behaviour [2]. But I don't really see a quick work around for that :( Which is a pitty, because having consistent behaviour would be really nice. Postgres is just sort of broken in that respect (although for a reason).
> 
> #### 2 - Different behaviour for literals and Series
> The special case in `__eq__` only considers when `other` is a python float with value `nan`, but if `other` is a Series that has the value `nan`, then the default behaviour happens. I think it's weird that the same comparison then gives a different result.
> 
> ```python
> df['nan'] = float('nan')
> df['eq_nan1'] = df['a'] == float('nan')
> df['eq_nan2'] = df['a'] == df['nan']
> # now 'eq_nan1' and 'eq_nan2' should have the same value, but they can be different
> ```
> 
> This is not a super big deal, but not ideal.
> 
> ### References
> [1] https://en.wikipedia.org/wiki/NaN#Comparison_with_NaN - NaN behaviour in IEEE 754 according to wikipedia, I didn't read the actual standard of course :)
> 
> [2] https://www.postgresql.org/docs/current/datatype-numeric.html#:~:text=power%20yields%20one.)-,Note,NaN%20values%20as%20equal%2C%20and%20greater%20than%20all%20non%2DNaN%20values.,-The%20types%20decimal

Removed the changes done to `__eq__`  and added instead `is_nan({})` in `Series.isnull`. Though this does not fixes the following situation:

```sql
WITH "loaded_data___c7a64e8d90f078b2ec409e1aa8486fde" AS (
        SELECT *
          FROM (VALUES FALSE, TRUE, TRUE, FALSE, TRUE)
            AS t("is_conversion_event")
       ),

    "fcsn" as (select CASE WHEN "is_conversion_event" THEN 1 ELSE CASE WHEN NOT "is_conversion_event" THEN 2 END END
                          AS "_first_conversion_step_number"
               from "loaded_data___c7a64e8d90f078b2ec409e1aa8486fde")

SELECT * FROM fcsn WHERE "_first_conversion_step_number" = cast(nan() as double)

```

The query from above breaks for Athena, due to `"_first_conversion_step_number" = cast(nan() as double)`. What I managed to understand is that since `_first_conversion_step_number` might have a nullable/infinite value, it causes a conflict when comparin with nan

thijs-obj commented on bach/bach/series/series.py:
```diff
@@ -1075,7 +1075,7 @@ def __getitem__(self: SeriesSubType, key: Union[Any, slice]) -> SeriesSubType:
     def isnull(self) -> 'SeriesBoolean':
         """
-        Evaluate for every row in this series whether the value is missing or NULL.
+        Evaluate for every row in this series whether the value is missing (NaN, None and SQL NULL).
```
There is some inconsistency between this function and `fillna` too.

Just thinking out loud here: Can we not just stick to only treating NULL as null in `fillna`, `dropna`, `isnull`, `notnull`, etc? Or do we actually have use-cases where we want to detect `NaN` too?

KathiaBarahona replied:
oops, I entirely forgot of `notnull`. Anyways, I considered adding `NaN` in the main implementation just for following pandas compatibility https://pandas.pydata.org/docs/reference/api/pandas.isnull.html

thijs-obj replied:
I think we could just stick to detecting `NULL`. We state in our core-concepts that that is one of the differences between our implementation and pandas (https://objectiv.io/docs/modeling/bach/core-concepts/#compatibility-with-pandas). But that is totally my opinion, maybe good for a wider discussion in the dev meeting on Monday.

If do we want to support `NaN` here, then I think we should be consistent and treat `NaN` the same in all functions (fillna, dropna, notnull, etc.)

KathiaBarahona replied:
I ended up moving this changes just for `dropna`, since we would tackle the main issue in a seperate PR

thijs-obj commented on bach/bach/series/series.py:
```diff
@@ -1085,12 +1085,18 @@ def isnull(self) -> 'SeriesBoolean':
         from bach import SeriesBoolean
+        expression = NonAtomicExpression.construct('{} is null', self)
+
+        if isinstance(self, SeriesAbstractNumeric):
```
Q: Why `SeriesAbstractNumeric` and not `SeriesFloat64`? An integer can never be `NaN`, right?

I think it is nicer to override this function in `SeriesFloat64` (or in `SeriesAbstractNumeric`) instead of checking the type of `self` here, because I think that's more consistent with how we do this in general.

KathiaBarahona replied:
In theory no, but maybe the result of the SeriesInt64 expression might yield a  NaN without Bach knowing

thijs-obj reviewed (commented):
some comments

thijs-obj commented:
> Removed the changes done to `__eq__` and added instead `is_nan({})` in `Series.isnull`. Though this does not fixes the following situation:
> 
> ```sql
> WITH "loaded_data___c7a64e8d90f078b2ec409e1aa8486fde" AS (
>         SELECT *
>           FROM (VALUES FALSE, TRUE, TRUE, FALSE, TRUE)
>             AS t("is_conversion_event")
>        ),
> 
>     "fcsn" as (select CASE WHEN "is_conversion_event" THEN 1 ELSE CASE WHEN NOT "is_conversion_event" THEN 2 END END
>                           AS "_first_conversion_step_number"
>                from "loaded_data___c7a64e8d90f078b2ec409e1aa8486fde")
> 
> SELECT * FROM fcsn WHERE "_first_conversion_step_number" = cast(nan() as double)
> ```
> 
> The query from above breaks for Athena, due to `"_first_conversion_step_number" = cast(nan() as double)`. What I managed to understand is that since `_first_conversion_step_number` might have a nullable/infinite value, it causes a conflict when comparin with nan

Nicely isolated problem like this. I think this is a bug in Athena :thinking: Very unfortunate, but doesn't seem like too big of a problem for us. Only get this error when  "_first_conversion_step_number" = cast(nan() as double) is in the WHERE clause, not if it is selected, so that's why I am thinking this is a bug.
